File: src/Array/ZigZagConversion/ZigZagConversion.java
class Solution {
    public String convert(String s, int numRows) {
        //Define StringBuilders
        StringBuilder[] sbs = new StringBuilder[numRows];
        for(int i = 0; i < numRows; i++){
            sbs[i] = new StringBuilder();
        }
        
        //Define Variables
        char[] arr = s.toCharArray();
        int n = arr.length;
        int index = 0;
        
        //Traverse zig zag
        while(index < n){
            //Go down
            for(int j = 0; j < numRows && index < n; j++){
                sbs[j].append(arr[index++]);
            }
            //Go Up before start
            for(int j = numRows - 2; j > 0 && index < n; j--){
                sbs[j].append(arr[index++]);
            }
        }
        
        //Combine all stringbuilders into one
        StringBuilder res = sbs[0];
        for(int i = 1; i < numRows; i++){
            res.append(sbs[i].toString());
        }
        return res.toString();
    }
}


/**
P     A     H    N
 A  P  L  S  I  I  G
   Y     I     R
**/

File: src/Array/ZigZagConversion/ZigZagConversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        # 1. Create n number of lists
        lists = []
        for i in range(numRows):
            lists.append([])

        # 2. Traverse the string in zig zag order
        index = 0
        while index < len(s):
            # Go down
            for i in range(numRows):
                if index == len(s):
                    break
                lists[i].append(s[index])
                index += 1

            # Go up
            for i in range(numRows - 2, 0, -1):
                if index == len(s):
                    break
                lists[i].append(s[index])
                index += 1
        
        # 3. Combine the lists into one str
        res = ""
        for cur_list in lists:
            res += "".join(cur_list)
        
        return res
